### ISSUE TITLE
ValueFromRemainingArguments parameters are not reflected as such in syntax diagrams

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#4626]
<sup>Related: #4715 and #4135.</sup>

It is currently **hard to discover if a given cmdlet/function has a `ValueFromRemainingArguments` parameter**, so I suggest we add a _new representation_ to syntax diagrams to signal that a given cmdlet / advanced supports does so.

We'll have to decide what an adequate representation is.

For instance, `Write-Host -?` shows the following for its `ValueFromRemainingArguments
` `-Object` parameter:

```powershell
[[-Object] <Object>]
```
which does not reflect its "specialness" - you can't tell that this parameter enables specifying multiple, independent arguments that are all bound to `-Object`, enabling convenient free-form invocations as `Write-Host a b c` in lieu of (the also supported) `Write-Host -Object a, b, c`

Perhaps something along the lines of (_updated based on the discussion below_):

```powershell
[-Object <Object> | <Object> ... ]
```

With an array-typed parameter, such as `Join-Path`'s `-AdditionalChildPath`:

```powershell
[-AdditionalChildPath <string[]> | <string> ...]
```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.5 on macOS 10.12.6
PowerShell Core v6.0.0-beta.5 on Ubuntu 16.04.3 LTS
PowerShell Core v6.0.0-beta.5 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
Windows PowerShell v5.1.15063.483 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
```

**Comment by iSazonov** [comment]
I can not repo this on Windows PowerShell v5.1.16257.1 (and 10240 too)

**Comment by mklement0** [comment]
@iSazonov: 

Sorry - I wasn't clear in my original title and post - please see if the update helps.

(I was also originally distracted by `Join-Path -?` not showing the new `-AddtionalChildPath` parameter at all - see https://github.com/PowerShell/PowerShell-Docs/issues/1612)

**Comment by iSazonov** [comment]
`[[-Object] <Object>[, <Object> ...] ]`

**Comment by mklement0** [comment]
Well, the `,` is tricky, because we want to convey that you do _not_ need it when taking advantage of the feature.

**Comment by iSazonov** [comment]
` [[-Object] <Object>[[,] <Object> ...] ] `

**Comment by mklement0** [comment]
Keep in mind that the parameter type need not be `<Object>`.

Among the core cmdlets that have `ValueFromRemainingArguments
` parameters, you'll also see `<string[]>`, for instance (`Get-Command -Syntax Join-Path`, parameter `-AdditionalChildPath`).

Conceptually, a `ValueFromRemainingArguments` is inherently a collection, given that it can bind multiple arguments.

The tricky part is that if you don't use the parameter name explicitly, you actually _must_ use individual arguments as opposed to an array to get the equivalent functionality
(at least with advanced functions (as opposed to compiled cmdlets), which brings us back to our perennial favorite, #2035).

In the example above, you'd really need two distinct representations:

`[-AdditionalChildPath <string[]>]`

or, without parameter name:

`[<string> ...]`

So perhaps the following?

```powershell
[-AdditionalChildPath <string[]> | <string> ...]
```

**Comment by iSazonov** [comment]
If #2035 will be fixed  comma should be here too?  ` [<string>, <string> ...]`

**Comment by mklement0** [comment]
[As I've argued here](https://github.com/PowerShell/PowerShell/issues/2035#issuecomment-323641345), I strongly feel that it should be fixed in the way that advanced functions already work (read: parameter binding must be fixed for _cmdlets_ - the _opposite_ of what #2035 proposes), which means that in no-parameter-name  invocations you _must not_ use `,` -  because use of `,` changes the semantics to sticking an _array_ into the _first element_ of the parameter variable.

**Comment by iSazonov** [comment]
Thanks for clarify! So we blocked until #2035 will be resolved.

**Comment by mklement0** [comment]
Not necessarily, because the `<string> ...` (comma-less) syntax will work either way, irrespective of how #2035 is resolved (incidentally, I just tried to [clarify my arguments](https://github.com/PowerShell/PowerShell/issues/2035#issuecomment-323816221) there).

The `<string> ...` (comma-less) syntax is not only the more convenient syntax, it is also the only one that works when you call a PS script from the outside.

In short: Either way, the comma-less syntax is the one worth highlighting in the syntax diagram.

**Comment by mklement0** [comment]
P.S.: To put it differently: You don't need `ValueFromRemainingArguments` at all for the `<string>, <string>, ...` syntax - just a regular positional parameter that accepts an array.

**Comment by iSazonov** [comment]
>the comma-less syntax is the one worth highlighting in the syntax diagram.

Thanks! Clear. It is good `[-AdditionalChildPath <string[]> | <string> ...]`

**Comment by cpdaatc** [comment]
https://github.com/PowerShell/PowerShell/iss



ues/4626
